### ISSUE TITLE
Is there any difference between this library and nanoemoji?

### ISSUE THREAD

**Issue by yisibl** [C1710/emoji_builder#7]
Nanoemoji is written in Python, and the performance of emoji_builder should be better. Do you have an idea to rewrite nanoemoji in Rust? I think someone at Google might sponsor you.

**Comment by C1710** [comment]
I have actually never heard of that tool, so I can't say where the different is...

**Comment by C1710** [comment]
As far as I understand it (from the Readme), it already uses Skia (or resvg, although I think that's only for reference) for rendering, so `emoji_builder` shouldn't be much or any faster than it, but I don't know.

**Comment by yisibl** [comment]
Yes, it uses skia to process SVG files and relies on https://github.com/googlefonts/picosvg.

What are the main things emoji_builder currently uses rust to handle?

**Comment by C1710** [comment]
> Yes, it uses skia to process SVG files and relies on https://github.com/googlefonts/picosvg.
> 
> What are the main things emoji_builder currently uses rust to handle?

So, right now, the process looks as follows:
1. Information about emojis (through Unicode's `emoji-test.txt`, etc.) is collected and combined with the SVG files that are present, creating structures that represent each emoji (this is done in I think like under 1 second or so), with some output on missing emojis
2. The builder (right now there's only one for CBDT/CBLC) checks whether the SVG source files have changed (by comparing their hashes to a list, which is much more robust to the timestamps, Make uses)
3. It then preprocesses the emojjs (currently this is just aligning the colors to a predefined color palette)
4. It renders the emojis using `resvg` (which uses `tinyskia`, which is a Rust-version of Skia, although without GPU support and therefore probably slower)
5. It adds that wavy style to the flag emojis
6. It then directly optimizes them with `oxipng`. All these steps are done in memory, all in parallel in Rust.
7. These emojis are then saved as files
8. Then the emojis are handed over to a subset of the scripts from Noto-Emoji, which are still written in Python, to build the actual font (on my not-too-bad PC, this takes like 3-4 seconds). I could remove some smaller parts with information collected through the processing the software has done before, but that's not very much.

So, the heaviest part is obviously rendering and optimizing. And (depending on the kind of color font they produce), I believe, nanoemoji will probably have a similar, if not better performance in that part by using the real Skia. Maybe the largest advantage here is that preprocessing, rendering, flag-waving and optimizing is done without first storing everything as files, but I don't know how nanoemoji does it. This is just a comparison to the scripts/Makefile that is used in the regular Noto Emoji repository.
